### ISSUE TITLE
vlc video audio

### ISSUE THREAD

**Issue by passenger94** [Shoes3/shoes3#133]
made a "video" branch on my repository

working for me using system libvlc 2.2.0
should work on Windows too according to vlc doc with a method change from **libvlc_media_player_set_xwindow** to **libvlc_media_player_set_hwnd**, there's also a cocoa method : **libvlc_media_player_set_nsobject** ...

tested with *simple-video.rb* with a remote and a local file

**Comment by ccoupe** [comment]
VLC has distribution problems for packaged Shoes - which libs to include in the package and
where to put them on the users system. Very difficult issue when it gets to codecs which may not be FOSS. That's the main reason it was ripped out of Shoes long ago.

**Comment by passenger94** [comment]
I was suspecting something like that ... reason why i used installed vlc,
could be set as an option though, like it was, except it could possibly work :-) let the user decide if it can/could deal with the distribution problems.
Of course it may not worth the maintenance work, though it was rather easy to rebuild the feature, i was missing the ability to have an easy solution for sounds inside shoes (i know there is gems for that but vlc is quite an amazing tool, there's also mpv but i found it to be more difficult to install headers)

**Comment by ccoupe** [comment]
For audio, the portaudio gem/extension source is still in Shoes - it's just not compiled and the libs aren't copied into the Shoes place. PITA if you get the wrong version of portaudio installed with apt-get.  VLC video also had threading problems with 1.9.1 (many things did) and given the distribution issues (and window issues at the time) no one wanted to rewrite the Shoes C code to deal with the new VLC api. I

**Comment by passenger94** [comment]
just to let you know it works nice here, made a video_gtk3 branch which is current with master (main point of this message) so if you want to foresee merging issues about gtk3 branch into master ...
(result + history)
cheers

**Comment by ccoupe** [comment]
Considering that the work that is going into #54 is based on the old Video code we should probably probably put the code back into Shoes.

**Comment by ccoupe** [comment]
I don't know how practical the following idea is. I don't want to build/maintain Shoes dependency for  libVLC (and all it's components) and I certainly don't want to bloat the Shoes download size by including it. However, we could  include  a fiddle/ffi based Ruby interface to the libVLC api and try to load the so/dylib/dll from known locations ( in cache.rb or shoes.rb ) if It loads and the version numbers are good enough then enable creation of Shoes Video Viideo widgets for Shoes scripts

VLC still has an external display window option. Obviously getting that done in Shoes is a bit of a challenge after all these years of neglect. Less clear is how to handle Control buttons like start/stop/pause/ff/rew and associated callbacks.

**Comment by ccoupe** [comment]
@passenger94 - would you create a video branch on  the  gitthub site and push to it so I can look at what you've done already.

**Comment by passenger94** [comment]
done, pending pull request on video_vlc branch  (still working by my side)

**Comment by ccoupe** [comment]
@passenger94. Thank you. I like that most of the VLC code is isolated into shoes/video/ instead of cluttering up ruby.c  Looking at your code has also helped me to think about my proposed fiddle/ffi method of dealing with Video.  In that **magical universe** , the shoes/video/* vlc code would be replaced with C that calls Ruby (using rb_eval) and the ruby code would use a dynamic  ffi/fiddle built bindings to talk to the vlc library - assuming the user has installed vlc on their system.   Some methods that deal with general Shoes things (top, left, move) would still be C. Any of your methods (for example) that call into vlc_apis would have to cross the Shoes.C- > Ruby/ffi -> vlc C  barriers.

The goal is to not to use C to compile vlc calls and thereby create a link dependency for vlc libs to be includes with Shoes. Side effect - you wouldn't need VLC installed to build a distributable Shoes that could handle VLC if installed on the users system. It's a lot to think about. I have no idea if Gtk3/Windows shoes_native_surface() returns something acceptable to vlc.

**Comment by passenger94** [comment]
:-D Now, That's what i'm calling Fiddling !  MAN !!! this is heavy stuff !!! Lots of fun :-D
arid  documentation + almost bare bones programming !!! ouch ! and i' have only discovered the beginning of  fiddle -> libvlc side, that is a pledge to be quite a funky experience when trying to render on a cairo surface !

**Comment by ccoupe** [comment]
Its very bare bones, learn as you go. Thankfully you have working code so it could be done one vlc call at a time until the last thing to do is delete `#include "vlc.h"`  Getting the structs properly configured is important and make sure nothing is lost in conversion from C into Ruby.  There are a couple of Projects  the have done ffi&vlc in Node.js and Scheme.   I wouldn't try Cairo drawing/manipulation/transforms this way but we need those do we for video widgets? A lot of work.

**Comment by passenger94** [comment]
ok, Thanks to the github goldmine, having a vlc instance playing a video/audio in an external window is quite straightforward, that's one thing, but 
interacting with structures and unpacking what's pointers are pointing to ........
there are some limitations apparently too, 
for example a union inside structure is not allowed in fiddle, we have to use a workaround for this.... create a place-holder and then squirm around fiddle pointers and sisters...
lots to learn :-D

**Comment by passenger94** [comment]
![shoe_fiddle_vlc](https://cloud.githubusercontent.com/assets/10730930/12521461/114c6864-c14b-11e5-990d-c27284a6afc3.jpg)


something is working :-)
fiddle + libvlc on the ruby side of Shoes
nothing in shoes C specific to vlc, generic api, but uses Shoes placements routines (here with margins)

**Comment by ccoupe** [comment]
Amazing!  Wonderful!

**Comment by ccoupe** [comment]
I have some head scratching ahead for Windows,  It kind of works, I get the audio track for an mp4
but error messages on the dos commandline and no video, 
```
E:\shoes3\samples>cshoes.exe videoFiddleVlc01.rb
[03da64c4] core video output error: video output creation failed
[03d84d3c] core decoder error: failed to create video output
[03da64c4] core video output error: video output creation failed
[03d84d3c] core decoder error: failed to create video output
```
VLC (standalone app) does show the video, so it's definitely s Shoes/Window problem. Probably the gdk call. to get a Handle.

**Comment by passenger94** [comment]
the libvlc hook to the drawing area is not yet totally finished (still missing some proper thickness to the widget to act like other widgets) but should work ... head scratching ... any chance to capture an actually helpful message ?
There was some complain about h264 (mp4) rendering ...

**Comment by ccoupe** [comment]
Yes, I've set the hook for Windows. Different error message but we have video. It's got new oddities but I'm convinced the fiddle concept you implemented will work. Even with the Windows issues, this will get into Shoes 3.3.1, as proof of concept, samples/very-expert-video.rb. or a widget.  I need to get the OSX code working at least as good as Window. Tomorrow, I hope.

**Comment by passenger94** [comment]
what do i need to get quickly set, on windows 7, to be able to mess around with c code ?

**Comment by ccoupe** [comment]
You will need RubyInstaller and the matching devkit. That should get you 
the MingW msys environment which is a kind of like the unix shell.  It's 
important to get the 32 bit versions of both.  So Ruby 2.2.3 is their 
latest and the 32 bit devkit for it.

http://rubyinstaller.org/downloads/

You'll probably want 7zip utilitiy so you can untar things from Shoe 
website. Plus some patience. The terminal is pretty lame re: 
cut/paste/select.

**Comment by BackOrder** [comment]
Is this part of 3.3.1 release? Some issues might need tagging.

**Comment by ccoupe** [comment]
Actually it's not tagged because it's not completely certain it's going to work or when it's going to work well enough. I think it's pretty close but I'm an optimist about schedules and features. We all know feature schedules are moveable.

**Comment by ccoupe** [comment]
The OSX code is going to take awhile. The new shoes_native_surface-* functions  are  a lot different from what cocoa.m has.  I think the newer code can be done in Cocoa once I get back in that head space.

**Comment by passenger94** [comment]
@ccoupe , Is Windows part working now ?
Looking at cocoa, makes me wondering if video was ever an option on Shoes osx ? The ancient code doesn't make much sense, it's missing a proper canvas to draw on, the equivalent of a GtkDrawingArea , at the moment "shoes_native_surface_new" returns the entire app canvas ...
Apart from that, it should be relatively straightforward (from my totally objective-C blind perspective, that is)

**Comment by ccoupe** [comment]
It should be just a case of creating an NSView with a NRect from the width and height and enablling/disabling certain event callbacks . I agree that VLC may have never worked on OSX.

The Windows code is still misbehaving. I got tired of working on Windows and decided to give the OSX codebase  some  attention.

**Comment by passenger94** [comment]
i'm concerned with this :
https://github.com/passenger94/shoes3/blob/video_fiddle_vlc/lib/shoes/videoffi.rb#L256

it works nice and we HAVE to wait for parent canvas to be drawn in order to accesss the drawable for vlc
But it might conflict with a user supplied start event in a shoes script
The only other solution, i found, is something like that but it's ugly... any idea ?
```ruby
 Thread.new do
        sleep 0.5
        drID = @video.drawable  # xlib window / HWND / NSView  id
        if libvlc_media_player_get_xwindow(@player) == 0
            libvlc_media_player_set_xwindow(@player, drID)
                # libvlc_media_player_set_hwnd       on Windows
                # libvlc_media_player_set_nsobject   on osx
         end
            
        play if @loaded && @autoplay
   end
```
Also waiting in a loop for app.started? (in a Thread) but it's not enough when app is started, asynchronous drawing is not yet finished ...

**Comment by ccoupe** [comment]
@passenger94 - sorry I didn't get back to you sooner.  I'm OK with adding undocumented/private methods.  We've got draw methods that should never be called from general purpose scripts.

**Comment by passenger94** [comment]
No problem :) 

On windows rake is complaining about not finding "libmsvcrt-ruby220.so", i thought it was looking for a dll !

**Comment by ccoupe** [comment]
Sounds like your Ruby install is OK.  Shoes is more complicated.  See https://github.com/Shoes3/shoes3/wiki/Building-Shoes-on-Windows

1) you need the precompiled dependencies,  (http://shoes.mvmanila.com/public/shoesdeps/mingw/ShoesDeps.zip - 96MB)
2) you need a win7-custom.yaml  - see https://github.com/Shoes3/shoes3/wiki/Custom.yaml. 
with your settings for your system. 3) Patience - it does work.

**Comment by passenger94** [comment]
yeah i followed that route, after remembering you took a lot of time to write Wiki pages about it ... !!
i'm sure it does work but the whole thing is kind of frightening to say the least ...

There is only a "libwinpthread.dll.a" inside ShoesDeps.zip, is it ok ? i was expecting a lot more (directories, dlls ...)
it seems like the win7-custom.yaml is called only if cross compiling, no ? i'm lost there
btw i'm trying to do a loose build, hoping it to be more easy, but i might be wrong probably because then i need all the dependencies installed , right ?

**Comment by ccoupe** [comment]
There is no loose build for Windows. They don't have sym links. So you have to do `rake win32:setup:win7` . Then you can rake, rake clean like normal.  win7-custom.yaml protects any changes you make to the rakefile from being blown away by a git refresh.  You want one. 

There should be lots of files in ShoesDep.zip, bin,lib include directories with things that make sense for unix folk.  It's always possible the file is bad.

**Comment by ccoupe** [comment]
Yes, the ShoesDep file is bad. I'll replace it.

**Comment by ccoupe** [comment]
The file is mingw.tar.gz . You'll need to use 7-Zip to unpacking.  Do NOT move the files into mingw/ruby directories. Whatever directory you unpack to belongs in the win7-custom.yaml 
```
Deps: E:/shoesdeps/mingw
Ruby: C:/Ruby22
Gemloc: E:/gems/shoes
Extloc: E:/gems/shoes
Exts:
  - ftsearch
  - chipmunk
Gems:
  - hpricot
  - sqlite3
InclGems:
  - sqlite3
  - nokogiri-1.6.7.1-x86-mingw32
  - mini_portile2-2.0.0
  - byebug-8.2.1
  - rb-readline-0.5.3
  - win32-shortcut-0.2.5
MS-Theme: true 
GtkVersion: 3
```
That's my win7-custom.yaml

**Comment by passenger94** [comment]
Thanks ! :-))

**Comment by ccoupe** [comment]
You'll also want to download the prebuilt gems unless you like to suffer.

**Comment by passenger94** [comment]
hmmmmmmm   .....  i think i might take the download route :-D , which is named ?

**Comment by ccoupe** [comment]
http://shoes.mvmanila.com/public/shoesdeps/prebuilt-gems/win7-gems.tar.gz

If you use a web browser you can navigate to all the files available for download. ;^)

**Comment by passenger94** [comment]
oh yes a web browser !!! ;-D   sorry

**Comment by passenger94** [comment]
ok things are moving forward,  But !
i have some "Package 'libpng' has no Name: field" which stops compilation, does it tell something to you ?
also the paths are hard-coded to your system in the .pc files, is it a problem ?

**Comment by ccoupe** [comment]
As I remember, @backOrder has used those deps (and .pc) files recently for 3.3.0 and I use them for both cross compiled (xwin7) and win7 builds which are pretty different.  There are several versions of pkg-config for windows,  however. I build from the dos console, Not from msys (and I think pkg-config is one of the reasons).

**Comment by ccoupe** [comment]
I just built win7 (master) myself.  Send me email ccoupe@cableon.net  if you want to work in a env where I send attachments.

**Comment by passenger94** [comment]
getting closer : cairo.pc wants **libpng16** not **libpng** on my system ...
now it almost compile, i'm having a video flag on the way, i think i need to git checkout to the right branch ...

**Comment by ccoupe** [comment]
Yes, it has not been updated for video (xwin7/env.rb) the changes for that branch. libpng.pc and libpng16.pc are duplicates.

**Comment by passenger94** [comment]
ok i'm there !
still not compiling though, because i need to figure out the video windows part now ...
looking in your branch where you have it i think

**Comment by passenger94** [comment]
that strange error will wait for tonight ! *./shoes/config.h:83:3: error: unknown type name 'HINSTANCE'*

**Comment by ccoupe** [comment]
That error suggests there are missing Windows .h files which should be in the devkit/msys. 

That gives me a few hours to deal with merge some branches. To really do some work on video for OSX, I want  some of the changes I've made in master. It does compile, and load the libvlc.dylib (which complains about no plugins) and then dies early in your fiddle setup.  

I'm going to try merging master into video_vlc branch. Expect a big commit.

**Comment by passenger94** [comment]
Compiles fine now, Thanks !
Hell, you sure reach level 1000000000000000+ in skills, tenacity and patience departments for managing all this madness !!!!!

**Comment by passenger94** [comment]
ok, it works for me, PR to test

**Comment by ccoupe** [comment]
Works for my Windows too! Congratulations! 

The rake files were a mess before I started 'fixing' them and they are still a mess.  One day when there is nothing to add to Shoes, we can clean them up. You do not want to look at what happens for cross compiling gems and exts.

**Comment by ccoupe** [comment]
OSX fun:  Had to resort to inserting raises into videoffi.rb:260 .  in initialize I did 
```ruby
        @vlci = libvlc_new(0, nil)
        @version = libvlc_get_version
	raise "vlc version#{@version} #{@vlci.inspect}"
        @player = libvlc_media_player_new(@vlci)
```
and the shoes console shows 
```
Error in <unknown> line 0 | 2016-02-07 23:17:42 -0700
vlc version2.2.1 Terry Pratchett (Weatherwax) #<Fiddle::Pointer:0x007fb780d13590 ptr=0x00000000000000 size=0 free=0x00000000000000>
/Users/ccoupe/build/xmavericks/Shoes.app/Contents/MacOS/lib/shoes/videoffi.rb:260:in `initialize'
```
If i remove the 'raise' then the next line, libvlc_media_player-new crashes Shoes (osx) something awful.  Linux returns
```
Error in <unknown> line 0 | 2016-02-07 23:57:39 -0700
vlc version2.1.6 Rincewind #<Fiddle::Pointer:0x000000029b2b50 ptr=0x0000000299e990 size=0 free=0x00000000000000>
/home/ccoupe/Projects/shoes3/dist/lib/shoes/videoffi.rb:260:in `initialize'
```
Windows (working) is also 2.2.1 so that can't be the problem.  Suspicious people might wonder if there is something wrong with the OSX vlc code.

**Comment by passenger94** [comment]
Nice ! (for windows) , ouchh !  i can barely imagine the cross compile hardcore poetry !

googling leads me to put some faith into believing that it possibly looks like osx might eventually need arguments to livlc_new, maybe try with this one : `https://github.com/strukturag/VLCKit/blob/master/Sources/VLCLibrary.m#L126`
or maybe ENV variable for plugins path   .....

**Comment by ccoupe** [comment]
Nice find for a a cocoa use of VLC. Very helpful. The online documentation appears to be wrong or incomplete. Who could have thought of that? :-)

**Comment by ccoupe** [comment]
cmd line args to libvlc_new() do print on the OSX launch terminal so that's nice. Still returns null from libvlc_new(). I suspect It's related to DYLD_LIBRARY_PATH and friends - see:
https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/dyld.1.html

I suspect that libvlc.dylib tries to load libvlccore.dyblib  using @loader_path and Shoes and fiddle are expecting @executable_path linkage. The causal OSX  audience may not have the `otool` and `libtool ` commands to rewrite, nor should that be done on a users system.  Assuming I'm correct, this going to be big problem and and at the moment I don't know how to test it other than copying the two vlc dylibs into Shoes.

**Comment by passenger94** [comment]
on linux this works :
``libvlc_new(2, ["--no-xlib"].pack('p1'))`
for 1 command line argument
`libvlc_new(2, ["--no-xlib", "--no-video-title-show"].pack('p2'))`
for 2 command line arguments
etc...

**Comment by ccoupe** [comment]
Pack? Of course. I knew there had to be a better way. Once the the args were passed in, vlc found and populated the plugins.dat file - another 50+  dylibs. First, I need to verify that DYLD_LIBRARY_PATH is the problem  (or solution)

**Comment by ccoupe** [comment]
Fortunately for all, it just takes setting the ENV['VLC_PLUGIN_PATH'] 
![osx-vid](https://cloud.githubusercontent.com/assets/222691/12967521/854c0b42-d024-11e5-9a46-781ea7512724.png)
Now I can undo all the ugly I added.

**Comment by BackOrder** [comment]
Awesome screenshot.

**Comment by ccoupe** [comment]
Awesome is the work of @passenger94 !!!

I do have a question for the international folks. We have hard coded paths to /Applications/VLC.app on OSX, for example. Is that a good path for internationalized OSX? . We check for "c:\Program Files (x86). Is that name used in all Windows? I have no idea.

Betas  for OSX and Windows at the usual place http://walkabout.mvmanila.com/public/shoes  Tight Linux builds have some issues which I know will be solvable. Documenting this ability for Shoes will be something of a challenge that I look forward too.  I can hardly wait to tell everyone! 

It's hard to describe how pleased I am.

**Comment by BackOrder** [comment]
> We have hard coded paths to /Applications/VLC.app on OSX, for example. Is that a good path for internationalized OSX?

My understanding is that translation is provided on UI level rather than filesystem level. The filesystem is all on the same on any localized version of MacOS X. Can anyone confirm this?

> We check for "c:\Program Files (x86). Is that name used in all Windows?

Actually, we are supposed to use %ProgramFiles% and %ProgramFiles(X86)%. Never hard coded path! Please, take note that translation is provided at UI level. Filesystem is in plain English.

**References**
https://technet.microsoft.com/en-us/library/cc749104(v=ws.10).aspx

**Comment by ccoupe** [comment]
Thanks @BackOrder , It appears that we have ENV['ProgramFiles(x86)'] already available (use samples/simple-info.rb) so that should be used . Here's  screen shot after that change
![win7-video](https://cloud.githubusercontent.com/assets/222691/12972356/bbafd5aa-d061-11e5-87d7-ba910221d192.png)

**Comment by BackOrder** [comment]
Take note that ENV['ProgramFiles(x86)'] is only available on 64 bit systems. 32 bit systems and older systems such as Windows XP use ENV['ProgramFiles'].

**Comment by passenger94** [comment]
YES ! cool, nice news !! :+1:

**Comment by ccoupe** [comment]
There are confounding issues with the case of not having vlc installed in Linux - it doesn't always trigger an exception from dlload(). That one I think I can fix (have fixed for Linux?)

There is a more difficult problem if the vlc version is too old.  Then videoffi.rb dies trying to extern something not in the older libvlc.  VLC 2.0 vs 2.1 and 2.2. ? You'd want to raise an exception that says "Your  version of VLC is too old"  I'm unclear where to do that. The @version string isn't available until 
import symbols is called. @passenger94 - Advice,  please

There's another error that may be related to my old linux chroot for x86 and i386 when run outside the chroot - They fail in ffi.so trying to load fiddle.so from Ruby which seems backwards to me. That could be that ruby wasn't compiled with ffi or fiddle or ...  Doesn't happen on the pi2 build which uses a more recent debian that the chroots. 

Lastly, there is @BackOrder mention of XP and 32 bit Windows. I have no way of testing on XP or interest in XP  and you'll probably run afoul of the older VLC version problem noted above. It's only a few lines of Ruby to add to your XP Shoes. If systems are too old to run the new video code (and I two linux chroots that won't) then it's time to update or do without video.

**Comment by passenger94** [comment]
for the libvlc version, we can 
- get rid of the `extern 'const char* libvlc_get_version()'` inside the import_symbols class method
- do the load lib voodoo, check only (no version checking) that libvlc loads (therefore exists - there is a Fiddle::DLError - )
- once loaded, before calling import symbols, declare `extern 'const char* libvlc_get_version()'` and check version with  libvlc_get_version()
- if everyone is happy, adult and consenting then import_symbols

sorry i didn't paid much attention to that, just thrown (threw ?) a minimal check 
also the Object monkey patching is temporary, and maybe some structure declaration not needed for the moment :
i'm still struggling with Fiddle to be able to check for width and height of a possible Video track, if someone have some advice ... ... i have different "working" (computer gods accepts the offering) routines but i get absurd numbers, and still unclear how to manage a union inside a structure in Fiddle context, which i need to access video, audio or subtitle track after calling `libvlc_media_tracks_get`
https://github.com/Shoes3/shoes3/blob/b667c0b21574cbce8502a786256b36ac453e8922/lib/shoes/videoffi.rb#L333

**Comment by ccoupe** [comment]
That's much like I was going to do for version checks but I thought I should ask you since you know more than I about fiddle and vlc. I'll constrain my changes to that part of videoffi.rb  FWIW,  one can get VLC 2.1.5 on old debians (7.2 timeframe) via the backport repo. Removing  it is not so obvious.  Shoes 3.3.1 works with that one on i686 and fails properly if there is no vlc installed.

**Comment by passenger94** [comment]
so it's under control now ? failing gracefully everywhere if problems ?

i managed, finally !, to get the width and height of the video track through `libvlc_media_tracks_get`using a void pointer in place of the union in the Media_track structure

**Comment by ccoupe** [comment]
Now it doing the right thing for  Linux, OSX & Windows when VLC is not installed or the version is too old. Also added a check for @BackOrder 32 bit XP "Program Files" but I can't test that.

I'm going to work on the manual for this feature. @passenger94 , if you get bored coding, you could always update the Wiki articles for installing Shoes on Windows or tell  me what instructions were wrong or outdated.

**Comment by passenger94** [comment]
Some food to think about the better protocol to call libvlc

i managed to compile latest git vlc (linux,  3.0.0 Vetinari) and our fiddling works !
I didn't installed it, so i gave the (not standard) path to libvlc to `Vlc.load_lib` AND i had to point the ENV['VLC_PLUGIN_PATH'] to the **modules** subdir ! (obviously becomes the plugins dir if installed)

So, at the moment, user have 2 different places to change in order to work with a random vlc lib, not good (and one is in shoes libs) ... also ENV['VLC_PLUGIN_PATH'] might be something to check on all platform (not standard installation is a frequent enough dandyism, i myself ....)
Thoughts ?

**Comment by passenger94** [comment]
i propose giving an optional "plugin_path" argument, set to a default nil, for **Vlc.load_lib** function
and inside that function after importing symbols setting the ENV['VLC_PLUGIN_PATH'] with the value if given.

**Comment by ccoupe** [comment]
I do think it's a good idea to allow for non standard locations (plural)  but rather than code that location into the users script,  why not another cobbler screen. Very similarly to how we can change the location of the packaging downloads.  Screen would allow user to enter the locations and create a small file in ~/.shoes/walkabout (and set two ENV['vars'] ). At Shoes startup, (lib/shoes/cache.rb) we can test for that file and if found, load the ENV['vars']  Should be pretty simple. 

we could also move/duplicate some of the code in vlc.load_lib to run at Shoes startup to populate the ENV[vars] if the file doesn't exist but the defaults paths do.

**Comment by passenger94** [comment]
oh yes, let's go to the cobbler, sounds good to me ! 
Beware, that guy is slowly but surely growing up ... ! ;-D

Speaking of back shop, i finally put the `libvlc_media_tracks_get` under control, after like decades of fiddle and C pointers bloody skirmish, (that was something !)

cleaning, synch and PR

**Comment by ccoupe** [comment]
I'll volunteer to update the cobbler, cache and and that part of videoffi.rb is you like.  I'm familiar with it. Unless you want to do it.

**Comment by passenger94** [comment]
please do so, i'm not familiar with the cobbler,  i should though !! :-D

**Comment by ccoupe** [comment]
I think I'm done with my mods.  Works well enough for me. I'd be happy for a  PR from @passenger94 and we can merge this branch into master and release 3.3.1. After documenting and testing again, of course.

**Comment by ccoupe** [comment]
>Edit: another advantage of non standard locations is that you can zip/pack libvlc with your script as an option and set paths accordingly (will be platform dependant, but it'll be a user choice only) 

Anybody that does that successfully cross platform  Will be named "King Of Shoes." All hail their suffering.

**Comment by passenger94** [comment]
haha ! i sure got that ! 
I mean if one wants to give a shoes script using video facility to someone else not having vlc or not wanting to install vlc, she/he could do it provided both being on same platform ! just send the (usual packed) script along with a copy of libvlc and plugins dir then set path accordingly, conceivable ? 
just tested ok with both libvlc and plugin dir in some random dir (250Mo+ though ...)
Your setup seems to work nice !!! Thanks !!!!!
Doing a PR with synched repo and further fixes, enhancements (regression?)

**Comment by ccoupe** [comment]
I've merged everything into master at github so that's were we should finish up and more changes to video.  I'm very pleased about this. Thank you for your hard work.

>do we need this ? :
https://github.com/Shoes3/shoes3/blob/video_vlc/samples/expert-video-player.rb#L113

I just felt it was a better example of how to determine that you have video capabilities and respond in a GUI friendly manner (the alert) rather than a blank screen with an error message in the hidden console and also moved the required and load into the Shoes.app {} block. Or demonstrating how to catch an expectation. 

> just send the (usual packed) script along with a copy of libvlc and plugins dir then set path accordingly, conceivable ? 
As long as you don't attempt to copy into directories you don't have permissions.   Clever people could create a gem with the tar balls inside and an extract script that runs when the gem is required and you could package that HUGE gem with your script. Even bigger GEM if it has all 5 architectures inside !

**Comment by passenger94** [comment]
something is missing here ?
https://github.com/Shoes3/shoes3/blob/master/shoes/video/video.c#L156

**Comment by ccoupe** [comment]
>something is missing here ?
https://github.com/Shoes3/shoes3/blob/master/shoes/video/video.c#L156

In this case, OSX is easier. Shoes_Control_Ref is * NSView which what vlc wants to draw in. No special casting is needed. 

>i have no way to test this, going from
https://github.com/Shoes3/shoes3/blob/master/lib/shoes/videoffi.rb#L263

In this case OSX is harder.  VLC can only be installed in /Applications and can't be moved from /Applications easily and if you do, it's not an OSX App that  you can browse in to find the libvlc.dylib and plugins. So, OSX can't have two VLC's, unless one of them is built from source but not installed.  Linux is actually similar.  Windows does allow VLC to be installed in different places so except for your situation (two vlc's, one not installed) only Windows really could use this feature, for regular folks.   I should remove that darwin condition however since it's not needed and interferes with the unlikely case that Shoes user actually has build from source vlc and a the normal VLC

**Comment by ccoupe** [comment]
Confirmed that on OSX, the Shoes ask_open dialogs will not go inside an App bundle. That's a good thing if you want to pick an App. Not so good to pick something inside an app.bundle. Current behavior is highly likely continue which means on OSX, VLC needs to be in /Applications.

**Comment by passenger94** [comment]
Ok ! noted :-)

**Comment by ccoupe** [comment]
I'm willing to close this issue and release 3.3.1 - Note, closing an issue doesn't mean it's 100% just that its good enough for people to use.
